samples/bluetooth/peripheral_hids_keyboard/src/main.c

```diff
@@ -0,0 +1,629 @@
+	0x13,	/* Key p */
+	0x16,	/* Key s */
+	0x12,	/* Key o */
+	0x09,	/* Key f */
+};
+
+static struct device		*led_dev;
+static struct device		*gpio_devs[4];
+static struct gpio_callback	gpio_cbs[4];
+static bool			caps_on;
+
+static struct conn_mode {
+	struct bt_conn *conn;
+	bool in_boot_mode;
+} conn_mode[CONFIG_BT_GATT_HIDS_MAX_CLIENT_COUNT];
+
+
+static void advertising_start(void)
+{
+	int err = bt_le_adv_start(BT_LE_ADV_CONN, ad, ARRAY_SIZE(ad),
+				  sd, ARRAY_SIZE(sd));
+
+	if (err) {
+		printk("Advertising failed to start (err %d)\n", err);
+		return;
+	}
+
+	printk("Advertising successfully started\n");
+}
+
+
+static void connected(struct bt_conn *conn, u8_t err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (err) {
+		printk("Failed to connect to %s (%u)\n", addr, err);
+		return;
+	}
+
+	printk("Connected %s\n", addr);
+
+	if (IS_ENABLED(CONFIG_BT_GATT_HIDS_SECURITY_ENABLED)) {
+		if (bt_conn_security(conn, sec_level)) {
+			printk("Failed to set security\n");
+		}
+	}
+
+	err = hids_notify_connected(&hids_obj, conn);
+
+	if (err) {
+		printk("Failed to notify HID service about connection\n");
+		return;
+	}
+
+	for (size_t i = 0; i < CONFIG_BT_GATT_HIDS_MAX_CLIENT_COUNT; i++) {
+		if (!conn_mode[i].conn) {
+			conn_mode[i].conn = conn;
+			conn_mode[i].in_boot_mode = false;
+			return;
+		}
+	}
+}
+
+
+static void disconnected(struct bt_conn *conn, u8_t reason)
+{
+	int err;
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	printk("Disconnected from %s (reason %u)\n", addr, reason);
+
+	err = hids_notify_disconnected(&hids_obj, conn);
+
+	if (err) {
+		printk("Failed to notify HID service about disconnection\n");
+	}
+
+	for (size_t i = 0; i < CONFIG_BT_GATT_HIDS_MAX_CLIENT_COUNT; i++) {
+		if (conn_mode[i].conn == conn) {
+			conn_mode[i].conn = NULL;
+			break;
+		}
+	}
+}
+
+
+static void security_changed(struct bt_conn *conn, bt_security_t level)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	printk("Security changed: %s level %u\n", addr, level);
+}
+
+
+static struct bt_conn_cb conn_callbacks = {
+	.connected = connected,
+	.disconnected = disconnected,
+	.security_changed = security_changed,
+};
+
+
+
+static void key_pack_queue(char *string, u8_t len)
+{
+	struct key_pack keys;
+	int err;
+
+	__ASSERT(len <= KEYS_MAX_LEN, "String is too long.");
+
+	printk("Getting key: %02X\n", (char) *string);
+
+	keys.elements = string;
+	keys.len = len;
+
+	err = k_msgq_put(&hids_queue, &keys, K_NO_WAIT);
+
+	if (err) {
+		printk("No space in the queue for button pressed\n");
+		return;
+	}
+}
+
+
+static void caps_lock_handler(const struct hids_rep *rep)
+{
+	u8_t report_val = (*rep->data) & OUTPUT_REPORT_BIT_MASK_CAPS_LOCK;
+
+	if (!caps_on && (report_val != 0)) {
+		gpio_pin_write(led_dev, CAPS_LOCK_LED, 0);
+
+		key_pack_queue(caps_on_key_scan_str,
```

Comment:
This could be removed to simplify the example

Reply:
I do not see real advantage from removing this 2 lines of the code. Does it really add any complexity?

Reply:
It adds code that is not wanted in a application. Setting a LED to indicate caps lock is, in my opinion, sufficient.

Reply:
This is direct copy of the HIDS keyboard example from the nRF5 SDK. If more people vote it that way Of course we can do it like that but it is a bit of nitpicking. Then I prefer to ask more people about theirs opinion.